tracks/julia/exercises/nucleotide-count/mentoring.md

```diff
@@ -92,4 +86,98 @@ function count_nucleotides4(strand)
+   This is faster because iterating characters from a `String` involves scanning the same vector of UTF-8 bytes but also doing a little work every iteration to establish if the current byte is the start of a multibyte sequence (and, if so to reassemble that sequence into its proper 32 bit character).
+2. The loop body is extremely simple and fast on x86 processors (load a value, increment a value at an offset\*) and branchless (there are no conditionals within the loop body).
+   Branchless code is often faster because branching hinders CPU instruction level parallelism and stops compilers using "Single Instruction Multiple Data" (SIMD) instructions (though this loop body can't benefit from SIMD anyway).
+3. Bounds checks are turned off with `@inbounds`, this removes some hidden branches (normally every array access in Julia has a little check that the index is valid first).
+   We can trivially prove that all indexes will be valid because a UInt8 has values ranging from 0:255 and the array we're indexing has indexes 1:256. The `+ 1` is there to make those two ranges line up.
+
+\* If you're good enough at reading assembly or LLVM you can verify this yourself with `@code_native` or `@code_llvm`. More info than you probably want:
+
+<details>
+<summary>More info than you want about assembly :D</summary>
+
+If you were writing the inner loop for `count_nucleotides4` manually in intel syntax x86 assembly you might write something like:
+
+```nasm
+        ; Where:
+        ; r14 is the address of the start of the utf8 array
+        ; rbx is the address of the start of the count array
+        ; rcx is the length of the utf8 array
+
+        ; Check utf8 isn't empty
+        cmp rcx rcx
+        je endofloop
+        ; init rdx to 0
+        xor rdx rdx
+countloop:
+        ; Get a byte from the utf8 array
+        movzx   edi, byte ptr [r14 + rdx + 8]
+        ; Increment the value in the count array at offset (byte * 8)
+        ; (* 8 because our array is of 64 bit ints = 8 bytes)
+        inc     qword ptr [rbx + 8*rdi]
+        ; Increment our index into the utf8 array
+        inc     rdx
+        ; If we haven't reached the end of the array yet, go again
+        cmp     rcx, rdx
+        jne     countloop
+endofloop:
+        ; ...
+```
+
+For our code above, Julia produces pretty similar results:
+
+(I used `code_native(count_nucleotides4, (String,); syntax=:intel)` which you could read as "Show me the native code produced when compiling `count_nucleotides4` with the argument types `(String,)` (i.e. one argument that is a `String`), and show me it in Intel's assembly syntax". You could also use `@code_native count_nucleotides4("foo")` and you'll get the same stuff but in the AT&T assembly format.)
+
+```nasm
+L320:
+        movzx   edx, byte ptr [r14 + rcx]
+        ; Load the address of counts (I don't know why this happens
+        ; repeatedly, but whatever).
+        mov     rdi, qword ptr [rax]
+        inc     qword ptr [rdi + 8*rdx]
```

Comment:
@bergjohan If you have some time, I'd appreciate it if you could consider this question briefly? Maybe there's an obvious answer that I am ignorant of!

Do you know if there's a reason for this bit `mov     rdi, qword ptr [rax]`. I think that's loading the address of `rax` into `rdi`, but I don't understand why it would be done inside a loop where it doesn't look like `rax` is getting modified?

Reply:
Wow, cool feature, being able to output assembly like this!

Yeah, I don't see any reason why we wouldn't be able to move the `mov rdi, qword ptr [rax]` line outside of the loop, instead of executing it in every iteration.

It seems like `rax` is a pointer to the `counts` array, so we're loading 8 bytes from the address that `rax` points to.
In C this would look like: https://godbolt.org/z/3344zv

Reply:
Yeah, it's a nice feature! Thanks for the second opinion <3